Add inline assembly pthread wrapper for aarch64

thiru-mcw opened the pull request:
This change introduces AArch64-specific inline assembly implementation that prepares the environment for starting a new thread. It performs the following actions:

- Loads the func pointer, arg, and stackbase from `wrapper_arg`
- Adjusts stack pointer with __CYGTLS_PADSIZE__ to ensure TLS alignment
- Reserves 32 bytes of shadow space & clears the frame pointer(x29)
- Moves the arg into x0 and calls the thread function via x12

thiru-mcw commented:
Validated by: https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/15251166764

thiru-mcw commented:
Also validated by running the `pthreads` tests from the internal `Cygwin test suite`.

**Pthreads Test results:**

```
+-----------+------+
| Result    | Count|
+-----------+------+
| PASS      |   72 |
| FAIL      |    2 |
| SKIP      |    0 |
| XFAIL     |    0 |
| XPASS     |    0 |
| ERROR     |    0 |
| TOTAL     |   74 |
+-----------+------+
```

**2 Failure tests:**

- FAIL: winsup.api/pthread/cancel11.exe
- FAIL: winsup.api/pthread/threadidafterfork.exe

Blackhex commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,22 @@ pthread_wrapper (PVOID arg)
-  // TODO
+  __asm__ __volatile__ (
+      "\n\
+    mov     x1, %[WRAPPER_ARG]  // Load &wrapper_arg into x1        \n\
```
Please keep the same indentation and alignment of `\n` consistent with x64 implementation.

thiru-mcw replied:
Sure, will fix now during rebase..

thiru-mcw replied:
Done & also rebase on top of latest `woarm64` branch.

Blackhex replied:
Now, it's OK. Thank you.

Blackhex commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,22 @@ pthread_wrapper (PVOID arg)
 	       [CYGTLS] "i" (__CYGTLS_PADSIZE__));
 #elif defined(__aarch64__)
-  // TODO
+  __asm__ __volatile__ (
```
Can please you explain why is `__volatile__` needed in this case? There does not seem to be any synchronization involved. E.g., can you give some example where the instructions re-ordering could cause the problem?

thiru-mcw replied:
Marking this asm block as `__volatile__ ` was my intention because it manipulates the stack pointer (sp) and calls a function via `blr x12`. I believe these actions must be preserved exactly as written to ensure the correct setup and invocation of the pthread start routine.

While most of the times the integrity of inline asm is preserved automatially by compiler, the `__volatile__` keyword again ensures the compiler does not optimize away or reorder this block.

In this case, removing `__volatile__` does not currently cause any test failures in pthreads, but it's included as a precaution to ensure thread safety during creation.
If a future change in optimization behaviour causes any regression, this `__volatile__` helps guard against breakages.

Blackhex replied:
OK, thank you.

Blackhex commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,22 @@ pthread_wrapper (PVOID arg)
+    blr     x12                 // Call thread func                 \n\
+    "
+      :
+      : [WRAPPER_ARG] "r"(&wrapper_arg), [CYGTLS] "i"(__CYGTLS_PADSIZE__)
```
Can you please explain what is behind the decision of using `r` constraint for `WRAPPER_ARG` instead of `o` as in x64 implementation?

thiru-mcw replied:
The `r` constraint is used for `WRAPPER_ARG` to ensure the address of `wrapper_arg` is placed in a general-purpose register. This is necessary because the AArch64 inline assembly relies on loading data from fixed offsets relative to that register. Using `r` ensures we have the base address in a register (x1), allowing multiple offset-based loads in the instruction sequence.

In x86_64, the `o` constraint is used to reference directly addressable memory operands, when the operand can be encoded directly as `[base + offset]` in a single instruction. While this works only in x64. aarch64 does not support varying offsets through "o".

Attempting to use `"o"` on AArch64 for `&wrapper_arg` leads to the following compile-time error: **_memory input 0 is not directly addressable_**
```
/home/crd/woarm/mingw-woarm64-build/code/cygwin/winsup/cygwin/create_posix_thread.cc:114:35: error: memory input 0 is not directly addressable
  114 |            : : [WRAPPER_ARG] "o" (&wrapper_arg),
      |                                   ^~~~~~~~~~~~
```

Blackhex replied:
Thank you for the explanation.

Blackhex commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,22 @@ pthread_wrapper (PVOID arg)
+    "
+      :
+      : [WRAPPER_ARG] "r"(&wrapper_arg), [CYGTLS] "i"(__CYGTLS_PADSIZE__)
+      : "x0", "x1", "x2", "x3", "x12", "x13", "x29", "memory");
```
Fixed.

thiru-mcw commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,20 @@ pthread_wrapper (PVOID arg)
+	   mov     x1, %[WRAPPER_ARG]  // Load &wrapper_arg into x1          \n\
+	   ldr     x12, [x1]           // Load thread func into x12          \n\
+	   ldr     x13, [x1, #8]       // Load thread arg into x13           \n\
+	   ldr     x2, [x1, #24]       // Load stackbase into x2             \n\
```
In wrapper_arg, stackbase is the `4th field` in the struct (offset 24 bytes).
https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/e475924ed35684d03495d996567651ffc85cd267/winsup/cygwin/create_posix_thread.cc#L28-L36
This also matches the x64 implementation.
`movq  24(%%rbx), %%rsp	# Load stackbase into rsp	\n\`

thiru-mcw commented:
Also Validated by: https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/15271338924

Blackhex commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,20 @@ pthread_wrapper (PVOID arg)
+	   mov     x3, %[CYGTLS]       // Load __CYGTLS_PADSIZE__ into x3    \n\
+	   sub     sp, x2, x3          // sp = stackbase - __CYGTLS_PADSIZE__\n\
+	   sub     sp, sp, #32         // Reserve 32 bytes (shadow space)    \n\
+	   mov     x29, xzr            // Clear frame pointer (x29 = 0)      \n\
```
Rather use `fp` for better readability.

thiru-mcw replied:
Sure, will fix in next push.

thiru-mcw replied:
fixed

Blackhex commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,20 @@ pthread_wrapper (PVOID arg)
+	   sub     sp, x2, x3          // sp = stackbase - __CYGTLS_PADSIZE__\n\
+	   sub     sp, sp, #32         // Reserve 32 bytes (shadow space)    \n\
+	   mov     x29, xzr            // Clear frame pointer (x29 = 0)      \n\
+	   mov     x0, x13             // Move thread arg into x0 (1st arg)  \n\
```
Releasing the original OS stack is missing in this implementation compared to x64. Isn't it needed?

thiru-mcw replied:
Thanks for highlighting the missing `VirtualFree`, the memory leak is now addressed. 

Initially, the thread function was stored in `x12`, a volatile register per the `AArch64` ABI. In `x86_64`, `r12` is non-volatile, so for strict parity, I’ve updated the implementation to use `x19` (for the thread function) and `x20` (for the thread argument), both of which are callee-saved on `AArch64`.

The logic now mirrors the x64 flow. After this change, test results remain consistent: **72/74 pthread tests pass**, with the same two failures (`cancel11`, `threadidafterfork`).

eukarpov replied:
the comment has been moved to another thread.

eukarpov commented on winsup/cygwin/create_posix_thread.cc:
```diff
@@ -100,7 +100,20 @@ pthread_wrapper (PVOID arg)
+	   sub     sp, sp, #32         // Reserve 32 bytes (shadow space)    \n\
+	   mov     x29, xzr            // Clear frame pointer (x29 = 0)      \n\
+	   mov     x0, x13             // Move thread arg into x0 (1st arg)  \n\
+	   blr     x12                 // Call thread func                   \n"
```
It might be optimized to something like this (not verified)

mov x9, %[WRAPPER_ARG]
ldr x10, [x9, #24]
sub sp, x10, %[CYGTLS]
sub sp, sp, #32
mov fp, xzr
ldp x9, x0, [x9]
blr x9

thiru-mcw replied:
Thanks for the suggestion @eukarpov, I agree with the suggested optimization it’s a better approach with minimal instructions.

for the current implementation, my primary goal was to maintain parity with the x64 code, which follows a similar structure. I feel aligning closely with the existing x64 logic helps ensure consistency and makes it easier to upstream the changes into the Cygwin codebase.

eukarpov replied:
I recommend writing efficient AArch64 code directly, rather than mapping an implementation from another architecture which is fine for prototyping.

thiru-mcw replied:
@eukarpov This optimization looks solid in terms of instruction reduction. but the assembler throws an immediate out of range error at line: `sub sp, x10, %[CYGTLS]`

The issue is that AArch64's `sub` instruction only supports `12-bit immediates` (max 0xFFF or 4095). Here, `%[CYGTLS]` is `0x3200` (12800), which can't be encoded directly.

So, here its worked around by splitting it into two instructions:
```
mov x11, %[CYGTLS]
sub sp, x10, x11
```

eukarpov replied:
Perhaps it might be also fixed by using
[CYGTLS] "r"(__CYGTLS_PADSIZE__)

thiru-mcw replied:
Thanks for the suggestions @eukarpov, they were very helpful!

I've incorporated the suggested optimizations, here's the updated code draft:

```
#elif defined(__aarch64__)
  __asm__ __volatile__ ("\n\
    mov     x9, %[WRAPPER_ARG]           // x9 = &wrapper_arg                     \n\
    ldr     x10, [x9, #24]               // x10 = wrapper_arg.stackbase           \n\
    sub     sp, x10, %[CYGTLS]           // sp = stackbase - CYGTLS_PADSIZE       \n\
    sub     sp, sp, #32                  // reserve shadow space                  \n\
    mov     fp, xzr                      // clear frame pointer                   \n\
    ldp     x19, x20, [x9]               // x19 = func, x20 = arg                 \n\
    mov     x0, sp                       // x0 = new stack pointer                \n\
    mov     x1, xzr                      // x1 = 0 (dwSize)                       \n\
    mov     x2, #0x8000                  // x2 = MEM_RELEASE                      \n\
    bl      VirtualFree                  // free original stack                   \n\
    mov     x0, x20                      // x0 = thread argument                  \n\
    blr     x19                          // call thread function                  \n"
    : : [WRAPPER_ARG] "r" (&wrapper_arg),
        [CYGTLS] "r" (__CYGTLS_PADSIZE__)
    : "x0", "x1", "x2", "x9", "x10", "x19", "x20", "x29", "x30", "memory");
#else
```

With this Implementation the final Cygwin test results remain unchanged (72/74 pthread tests passing).

Additionally, I had a brief discussion with the Cygwin maintainers. Their preference is to prioritize correctness and readability using idiomatic AArch64 assembly, accompanied by clear comments to explain the assembler logic. 
W.r.t Optimizations they are optional and should only be applied if backed by real benchmarks, rather than speculative performance gains.

eukarpov replied:
In current implementation It makes sense
to replace x9 with x19
remove
 ldp     x19, x20, [x9]
 
 and replace 
  mov x0, x20
  with
  ldp x19, x0, [x19]

eukarpov replied:
Just a note: 
Talking about further optimization, just out of interest 
sub     sp, x10, %[CYGTLS]
sub     sp, sp, #32
could be replaced with one command

sub sp, x10, %[CYGTLS], lsl 4"
[CYGTLS] "r" ((__CYGTLS_PADSIZE__ >> 4) + 2)

thiru-mcw replied:
Done, applied both suggestions and updated the PR.

thiru-mcw replied:
Validated in Local setup, CI check [In-Progress] https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/15392187902

eukarpov replied:
Thanks, LGTM. 
Perhaps x30 should not be mentioned in the list.

thiru-mcw replied:
fixed, removed LR from the list.

eukarpov replied:
for better readability after all as it is also applied I would change it to
sub sp, x10, %[CYGTLS]" \
:: [CYGTLS] "r" (__CYGTLS_PADSIZE__ + 32)

thiru-mcw replied:
Done, also added a comment to state `CYGTLS_PADSIZE + 32`
https://github.com/Windows-on-ARM-Experiments/newlib-cygwin/blob/0221d8fd3c81c31e9343f71e78e94fe8556304b5/winsup/cygwin/create_posix_thread.cc#L117

Blackhex reviewed (approved):
LGTM, approving preferring readability/maintanability over over-optimization as this is hardly to be a hot-spot.

thiru-mcw committed:
Aarch64: Add inline assembly pthread wrapper

This change introduces AArch64-specific inline assembly implementation
that prepares the environment for starting a new thread.

eukarpov reviewed (approved):
Thanks for addressing and verifying multiple review comments.
The final version looks good to me.

thiru-mcw commented:
The Test suite regression in the pthread_wrapper implemetation is fixed by https://github.com/cygwin/cygwin/commit/10380e9f446640c8d1bd4d60ec085bb31efa26f1.
